DOCSP-41148 Authentication Guides (#50)
mongodb/docs-kotlin-sync: source/security/authentication.txt

.. _kotlin-sync-auth:

=========================
Authentication Mechanisms
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: validate credentials, protocols, code example 

Overview
--------

In this guide, you can learn how to authenticate with MongoDB using each
**authentication mechanism** available in the MongoDB Community Edition.
Authentication mechanisms are processes by which the driver and server
confirm identity and establish trust to ensure security.

To select a specific authentication mechanism, you can specify the
mechanism, your credentials, and other necessary information
in the options of your connection string or in a ``Credential`` struct.

To learn more about the connection string options for authentication,

> see the :manual:`Authentication Options

Reviewer: Link should be: https://www.mongodb.com/docs/manual/reference/connection-string-options/#authentication-options

Author: Linked to https://www.mongodb.com/docs/manual/reference/connection-string/#connection-string-formats, which leads to helpful information on string format options as well as link to the above page^

Author: I'll remove the link entirely. Initially wanted to include because I ran into trouble with ssl being enabled depending on whether the srv or standard string format was used. I've since updated the tutorial samples, so hopefully that won't be an issue for users any more.

</reference/connection-string/#authentication-options>` section
of the Connection String URI Format guide in the Server manual.

To authenticate using ``Kerberos`` or ``LDAP``, see the
:ref:`Enterprise Authentication Mechanisms guide <kotlin-sync-enterprise-auth>`.

For more information on establishing a connection to your MongoDB cluster,
see :ref:`Connect to MongoDB <kotlin-sync-connect>`.

.. _kotlin-sybc-auth-default:

Default
-------

The default authentication mechanism setting uses one of the following
authentication mechanisms depending on what your MongoDB server supports:

#. ``SCRAM-SHA-256``
#. ``SCRAM-SHA-1``

Server versions 4.0 and later use ``SCRAM-SHA-256`` as the default
mechanism.

The following code snippets show how to use the default authentication mechanism,

> using the following placeholders:

Reviewer: I: [using](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#u:~:text=Don%27t%20use%20using%20(alone)%20or%20with.)

applies throughout

using the following placeholders:

* ``db_username`` - your MongoDB database username
* ``db_password`` - your MongoDB database user's password
* ``hostname`` - network address of your MongoDB server, accessible by your client
* ``port`` - port number of your MongoDB server
* ``authenticationDb`` - MongoDB database that contains your user's
  authentication data. If you omit this parameter, the driver uses the
  default value ``admin``.

Select the :guilabel:`Connection String` or the :guilabel:`MongoCredential`
tab below for instructions and sample code for specifying this authentication
mechanism:

.. tabs::

   .. tab::
      :tabid: Connection String

      To specify the default authentication mechanism by using a connection
      string, omit the mechanism. Your code to instantiate a ``MongoClient``
      should resemble the following:

>       .. literalinclude:: /includes/security/authentication.kt

Author: How do you usually concatenate? Also fine with leaving as is, as the scroll isn't too long.

         :language: kotlin
         :start-after: start-default-cred-string
         :end-before: end-default-cred-string
         :dedent:
      
   .. tab::
      :tabid: MongoCredential

      To specify the default authentication mechanism by using the
      ``MongoCredential`` class, use the ``createCredential()`` method.
      Also, enable TLS by calling the
      `applyToSslSettings() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.Builder.html#applyToSslSettings(com.mongodb.Block)>`__
      method and setting the ``enabled`` property to ``true`` in the

>       `SslSettings.Builder <{+api+}/apidocs/mongodb-driver-core/com/mongodb/connection/SslSettings.Builder.html>`__

Author: Fixed!

The ``MONGODB-AWS`` authentication mechanism uses your Amazon Web Services
Identity and Access Management (AWS IAM) credentials to authenticate your
user. To learn more about configuring MongoDB Atlas, see the
:atlas:`Set Up Authentication with AWS IAM </security/aws-iam-authentication>`
guide.

To instruct the driver to use this authentication mechanism, you can specify
``MONGODB-AWS`` either as a parameter in the connection string or by using
the ``MongoCredential.createAwsCredential()`` factory method.

Learn how to specify this authentication mechanism and the various ways to
provide your AWS IAM credentials in the next sections.

These sections contain code examples that use the following placeholders:

* ``awsKeyId`` - value of your AWS access key ID
* ``awsSecretKey`` - value of your AWS secret access key
* ``atlasUri`` - network address of your MongoDB Atlas deployment
* ``hostname`` - hostname of your MongoDB Atlas deployment
* ``port`` - port of your MongoDB Atlas deployment
* ``awsSessionToken`` - value of your AWS session token

.. _kotlin-mongodb-aws-sdk:

AWS SDK 
~~~~~~~

You can use one of the AWS SDK for Java v1 or v2 to specify your credentials.

> This method offers the following features:

Reviewer: ```suggestion
Using the AWS SDK offers the following features:
```

- Multiple options for obtaining credentials

> - Credential caching which helps your application avoid rate limiting

Reviewer: I: [which](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#std-term-that:~:text=Use%20which%20to%20introduce%20a%20nonrestrictive%20clause%2C%20and%20use%20a%20comma.)

```suggestion
- Credential caching, which helps your application avoid rate limiting
```

> - Credential provider management for use with the `Elastic Kubernetes Service <https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html>`__.

Reviewer: ```suggestion
- Credential provider management for use with the `Elastic Kubernetes Service. <https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html>`__
```

> To use the AWS SDK for Java for ``MONGODB-AWS`` authentication, you must

Reviewer: S: remove "for Java" to improve readability. (hopefully the user understands they're working in Java...)

```suggestion
To use the AWS SDK for ``MONGODB-AWS`` authentication, you must
```

> perform the following:

Reviewer: I: [following](https://www.mongodb.com/docs/meta/style-guide/style/lists/#:~:text=Don%E2%80%99t%20tell%20users%20to%20%22do%20the%20following.%22)

```suggestion
perform the following steps:
```

> 1. Specify the authentication mechanism
> #. Add the SDK as a dependency to your project
> #. Supply your credentials by using one of the methods in the credential
>    provider chain

Reviewer: I: [punctuation](https://www.mongodb.com/docs/meta/style-guide/style/lists/#:~:text=In%20a%20list%20of%20only%20sentences%2C%20including%20imperative%20statements%2C%20use%20punctuation%20at%20the%20end%20of%20each%20item.)

```suggestion
1. Specify the authentication mechanism.
#. Add the SDK as a dependency to your project.
#. Supply your credentials by using one of the methods in the credential
   provider chain.
```

> To specify the authentication mechanism by using a ``MongoCredential``,

Reviewer: ```suggestion
To specify the ``MONGODB-AWS`` authentication mechanism by using a ``MongoCredential`` object,
```

> use the ``MongoCredential.createAwsCredential()`` factory method

Reviewer: ```suggestion
call the ``MongoCredential.createAwsCredential()`` factory method
```

and add the ``MongoCredential`` instance to your ``MongoClient`` as shown
in the following example:

.. literalinclude:: /includes/security/authentication.kt
   :language: kotlin
   :dedent:
   :start-after:  start-aws-sdk-mcred
   :end-before: end-aws-sdk-mcred
   :emphasize-lines: 1, 9

> To specify the authentication mechanism in the connection string, add

Reviewer: ```suggestion
To specify the ``MONGODB-AWS`` authentication mechanism in the connection string, add
```

it as a parameter as shown in the following example:

.. literalinclude:: /includes/security/authentication.kt
   :language: kotlin
   :dedent:
   :start-after:  start-aws-sdk-cred-string
   :end-before: end-aws-sdk-cred-string

To add the AWS SDK as a dependency to your project, see the following

> AWS documentation for the version you need:

Reviewer: Q: Do you know when a user would need v1 vs. v2?

Author: Just depending on which java version the user wants to use. Looked into it more, and the java v1 version is reaching end of support next year. Included a note at the top of section to recommend using v2 instead.

  section.

- To learn more about the **AWS SDK for Java v1** class the driver uses to
  get the credentials, see the `DefaultAWSCredentialsProviderChain <https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html>`__
  API documentation.

  Learn how to supply your credentials to this class from the
  `Using the Default Credential Provider Chain <https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#credentials-default>`__
  section.

.. note::

   If you include both v1 and v2 of the AWS SDK for Java in your project,
   you must use the v2 methods to supply your credentials.

.. _kotlin-mongodb-aws-env-variables:

Specify Your Credentials in the Environment
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can provide your AWS IAM credentials by instructing the driver to
use the ``MONGODB-AWS`` authentication mechanism and by setting the
appropriate environment variables.

To use the environment variables to supply your credentials, you must perform
the following:

> 1. Specify the authentication mechanism
> #. Add the appropriate environment variables

Reviewer: ```suggestion
1. Specify the authentication mechanism.
#. Add the appropriate environment variables.
```

> You can specify the authentication mechanism by using a ``MongoCredential``
> or on the connection string.

Reviewer: ```suggestion
You can specify the ``MONGODB-AWS`` authentication mechanism by using a ``MongoCredential``
object or in the connection string.
```

> To specify the authentication mechanism by using a ``MongoCredential``,

Reviewer: ```suggestion
To specify the authentication mechanism by using a ``MongoCredential`` object,
```

> use the ``MongoCredential.createAwsCredential()`` factory method and add the

Reviewer: ```suggestion
call the ``MongoCredential.createAwsCredential()`` factory method and add the
```

``MongoCredential`` instance to your ``MongoClient`` as shown in the following
example:

.. literalinclude:: /includes/security/authentication.kt
   :language: kotlin
   :dedent:
   :start-after:  start-aws-env-mcred
   :end-before: end-aws-env-mcred
   :emphasize-lines: 1, 9

> To specify the authentication mechanism in the connection string, add it as a

Reviewer: ```suggestion
To specify the ``MONGODB-AWS`` authentication mechanism in the connection string, add it as a
```

The following example shows how you can set your **programmatic access keys**
in environment variables by using ``bash`` or a similar shell:

.. code-block:: bash

   export AWS_ACCESS_KEY_ID=<awsKeyId>
   export AWS_SECRET_ACCESS_KEY=<awsSecretKey>
   export AWS_SESSION_TOKEN=<awsSessionToken>

Omit the line containing ``AWS_SESSION_TOKEN`` if you don't need an AWS
session token for that role.

To authenticate by using **ECS container credentials**, set the ECS
endpoint relative URI in an environment variable by using ``bash`` or
a similar shell as shown in the following example:

.. code-block:: bash

   export AWS_CONTAINER_CREDENTIALS_RELATIVE_URI=<your ECS endpoint>

To authenticate by using **EC2 container credentials**, make sure none of the
aforementioned environment variables are set. The driver obtains the
credentials from the default IPv4 EC2 instance metadata endpoint.

.. _kotlin-mongodb-aws-mongoclient-configuration:

Specify Your Credentials in a MongoCredential
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can supply your AWS IAM credentials to a ``MongoClient`` by using a
``MongoCredential`` instance. To construct the ``MongoCredential`` instance

> for ``MONGODB-AWS`` authentication,  use the `createAwsCredential() <{+core-api+}/com/mongodb/MongoCredential.html#createAwsCredential(java.lang.String,char%5B%5D)>`__

Reviewer: ```suggestion
for ``MONGODB-AWS`` authentication,  call the `createAwsCredential() <{+core-api+}/com/mongodb/MongoCredential.html#createAwsCredential(java.lang.String,char%5B%5D)>`__
```

factory method.

You can supply only programmatic access keys to the
``MongoCredential.createAwsCredential()`` method. If you need to supply ECS

> or EC2 container credentials, use the instructions in

Reviewer: ```suggestion
or EC2 container credentials, follow the instructions in
```

:ref:`<kotlin-mongodb-aws-env-variables>` or :ref:`<kotlin-mongodb-aws-sdk>`.

> To use the ``MongoCredential`` for ``MONGODB-AWS`` authentication, you

Reviewer: ```suggestion
To use a ``MongoCredential`` object for ``MONGODB-AWS`` authentication, you
```

> must perform the following:
> 
> 1. Specify the authentication mechanism
> #. Supply the credentials

Reviewer: ```suggestion
must perform the following steps:

1. Specify the authentication mechanism.
#. Supply the credentials.
```

> To specify the authentication mechanism by using a ``MongoCredential``,

Reviewer: ```suggestion
To specify the authentication mechanism by using a ``MongoCredential`` object,
```

> use the ``MongoCredential.createAwsCredential()`` factory method

Reviewer: ```suggestion
call the ``MongoCredential.createAwsCredential()`` factory method
```

> and add the ``MongoCredential`` instance to your ``MongoClient`` as shown

Reviewer: ```suggestion
and add the ``MongoCredential`` instance to your ``MongoClient``, as shown
```

in the following example:

.. literalinclude:: /includes/security/authentication.kt
   :language: kotlin
   :dedent:
   :start-after:  start-aws-mcred
   :end-before: end-aws-mcred
   :emphasize-lines: 1, 9

If you need to specify an AWS session token, pass it to the
`withMechanismProperty() <{+core-api+}/com/mongodb/MongoCredential.html#withMechanismProperty(java.lang.String,T)>`__

> method as shown in the following example:

Reviewer: ```suggestion
method, as shown in the following example:
```

.. literalinclude:: /includes/security/authentication.kt
   :language: kotlin
   :dedent:
   :start-after:  start-aws-mcred-wmechprop
   :end-before: end-aws-mcred-wmechprop
   :emphasize-lines: 1, 2, 10

To refresh your credentials, you can declare a ``Supplier`` lambda expression

> that returns new credentials as shown in the following example:

Reviewer: ```suggestion
that returns new credentials, as shown in the following example:
```

.. literalinclude:: /includes/security/authentication.kt
   :language: kotlin
   :dedent:
   :start-after:  start-aws-lambda-expression
   :end-before: end-aws-lambda-expression
   :emphasize-lines: 4-6, 9

If you must provide AWS IAM credentials in a connection string, you can add 

> it to your ``MongoClientSettings`` by calling the `applyConnectionString() <{+core-api+}/com/mongodb/MongoClientSettings.Builder.html#applyConnectionString(com.mongodb.ConnectionString)>`__

Reviewer: ```suggestion
it to your ``MongoClientSettings`` object by calling the `applyConnectionString() <{+core-api+}/com/mongodb/MongoClientSettings.Builder.html#applyConnectionString(com.mongodb.ConnectionString)>`__
```

method:

.. literalinclude:: /includes/security/authentication.kt
   :language: kotlin
   :dedent:
   :start-after:  start-aws-apply-connect-string
   :end-before: end-aws-apply-connect-string
   :emphasize-lines: 2, 5

Additional Information
----------------------

To learn more about authenticating to MongoDB, see

> :manual:`Authentication </core/authentication/>` in the Server manual.

Reviewer: ```suggestion
:manual:`Authentication </core/authentication/>` in the {+mdb-server+} manual.
```

To learn more about managing users of your MongoDB deployment, see

> :manual:`Users </core/security-users/>` in the Server manual.

Reviewer: ```suggestion
:manual:`Users </core/security-users/>` in the {+mdb-server+} manual.
```

   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: validate credentials, protocols, code example 

Overview
--------

In this guide, you can learn how to authenticate with MongoDB using each
**authentication mechanism** available in the MongoDB Community Edition.
Authentication mechanisms are processes by which the driver and server
confirm identity and establish trust to ensure security.

This guide describes the following authentication mechanisms:

- :ref:`SCRAM-Based Mechanisms <kotlin-sync-auth-scram-mechanisms>`
- :ref:`MONGODB-AWS Mechanism <kotlin-sync-auth-aws>`
- :ref:`MONGODB-X509 Mechanism <kotlin-sync-auth-x509>`

To select a specific authentication mechanism, you can specify the
mechanism, your credentials, and other necessary information
in the options of your connection string or in a ``Credential`` struct.

In this guide, the examples demonstrate how to configure
authentication in a ``Credential`` struct.

To learn more about the connection string options for authentication,
see the :manual:`Authentication Options
</reference/connection-string/#authentication-options>` section

> of the Connection String URI Format guide in the Server manual.

Author: I think the info on the differences between the ``+srv`` and the standard form of the connection string is helpful ([Connection String Formats](https://www.mongodb.com/docs/manual/reference/connection-string/#connection-string-formats)). The page also contains info on how to find connection string, which isn't linked to anywhere else.

To authenticate using ``Kerberos`` or ``LDAP``, see the

> :ref:`Enterprise Authentication Mechanisms guide <kotlin-sync-enterprise-auth>`.

Author: Edited to a note, what do you think?

For more information on establishing a connection to your MongoDB cluster,
see :ref:`Connect to MongoDB <kotlin-sync-connect>`.

.. _kotlin-sync-auth-scram-mechanisms:

SCRAM-Based Mechanisms
----------------------

> Salted challenge response authentication mechanism (SCRAM) refers to a

Author: Removed this section

group of authentication mechanisms that use a username and
password to authenticate to a server.

MongoDB supports the following SCRAM-based authentication mechanisms:

- :ref:`SCRAM-SHA-256 <kotlin-sync-auth-scramsha256>`: an authentication mechanism that

>   uses your database username and password, encrypted with the ``SHA-256``

Author: Think "your" is fine, as it engages the reader and makes it more action-oriented (referencing the [Using Second Person entry](https://www.mongodb.com/docs/meta/style-guide/writing/write-to-the-user/) in style guide).

MONGODB-X509
------------

The ``MONGODB-X509`` authentication mechanism uses
:wikipedia:`TLS <Transport_Layer_Security>` with X.509 certificates to
authenticate your user. When you specify the ``X.509``
authentication mechanism, the server authenticates the connection by using
the subject name of the client certificate.

The following code snippets show how to specify the authentication mechanism 
by using the following placeholders:

> * ``hostname``: network address of your MongoDB deployment, accessible by your client.
> * ``port``: port number of your MongoDB server.
> * ``authenticationDb``: MongoDB database that contains your user's

Reviewer: ```suggestion
* ``hostname``: The network address of your MongoDB deployment, accessible by your client.
* ``port``: The port number of your MongoDB server.
* ``authenticationDb``: The MongoDB database that contains your user's
```

:atlas:`Set Up Authentication with AWS IAM </security/aws-iam-authentication>`
guide.

To instruct the driver to use this authentication mechanism, you can either
specify ``MONGODB-AWS`` as a parameter in the connection string or call
the ``MongoCredential.createAwsCredential()`` factory method.

In the following sections, you can learn different ways to specify the 
``MONGODB-AWS`` authentication mechanism and provide your AWS IAM credentials.

These sections contain code examples that use the following placeholders:

* ``awsKeyId``: value of your AWS access key ID
* ``awsSecretKey``: value of your AWS secret access key
* ``atlasUri``: network address of your MongoDB Atlas deployment
* ``hostname``: hostname of your MongoDB Atlas deployment
* ``port``: port of your MongoDB Atlas deployment

> * ``awsSessionToken``: value of your AWS session token

Reviewer: Start items with "The"

.. _kotlin-mongodb-aws-sdk:

AWS SDK 
~~~~~~~

.. note:: End of Support for AWS SDK for Java v1

>    The AWS SDK for Java v1 will reach end of support on Decemeber 31, 2025. 

Reviewer: ```suggestion
   The AWS SDK for Java v1 will reach end of support on December 31, 2025. 
```
